nft_ingester/src/program_transformers/bubblegum/db.rs

```diff
@@ -441,3 +597,216 @@ where
 
     Ok(())
 }
+
+pub async fn upsert_creators<T>(
+    txn: &T,
+    id: Vec<u8>,
+    creators: &Vec<Creator>,
+    slot_updated: i64,
+    seq: i64,
+    delete_existing: bool,
+) -> Result<(), IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    // Begin a transaction.  If the transaction goes out of scope (i.e. one of the executions has
+    // an error and this function returns it using the `?` operator), then the transaction is
+    // automatically rolled back.
+    let multi_txn = txn.begin().await?;
+
+    if creators_should_be_updated(&multi_txn, id.clone(), seq).await? {
+        if delete_existing {
+            // Delete any existing creators.
+            asset_creators::Entity::delete_many()
+                .filter(Condition::all().add(asset_creators::Column::AssetId.eq(id.clone())))
+                .exec(&multi_txn)
+                .await?;
+        }
+
+        if !creators.is_empty() {
+            // Vec to hold base creator information.
+            let mut db_creator_infos = Vec::with_capacity(creators.len());
+
+            // Vec to hold info on whether a creator is verified.  This info is protected by `seq` number.
+            let mut db_creator_verified_infos = Vec::with_capacity(creators.len());
```

Comment:
Let me look at this more, it still needs more work based on not overwriting creator_verify and creator_unverify instructions that happen after `mint_v1` or `update_metadata`.

Reply:
_So the point I am at with the creators table is this:_

## Problem statement
Consider indexing the following Bubblegum instructions:
* `mint_v1` (has creators A and B both unverified)
* `verify_creator` (creator A)
* `update_metadata` (remove all creators)

With out-of-order indexing, the `update_metadata` could be indexed first, resulting in no creators being in the `asset_creators` table for this asset.  Then if the `verify_creator` instruction was indexed next, it would attempt to re-insert creator A.  Because this creator was already removed by a higher-sequence-numbered instruction, it should not be inserted.

There isn't a value to check in the `asset_creators` table that indicates that `update_metadata` already happened.  My solution is to use the asset table's `base_info_seq` to protect against these incorrect creator insertions.  Also, I want to just use the normal `asset.seq` to prevent overwriting changes from account-based Token Metadata indexing after an asset was decompressed.  This will be better than using `slot_updated > asset.slot_updated` for indexing Bubblegum instructions, because `mint_v1` and `update_metadata` could be in the same slot, resulting in missed data.

The drawback of using `asset_table` values before making changes to the `asset_creators` table is that it opens us up to race conditions with concurrent operations.

## Possible solutions

1. Use serializable transactions to enforce serializable isolation level.
2. Use traditional database locking to prevent concurrent access to the tables being modified.

### Serializable transactions
#### Pros:
* High level of isolation ensuring that the results of concurrent transactions are equivalent to some serial execution.
* Easier to code because we just put all the affected SQL into serializable transactions and don't have to manage locks.

#### Cons:
* Perhaps to make it all work as expected and truly get the guarantees, we would have to put the access of the `asset` table and `asset_creators` tables in `mint_v1`, `update_metadata`, AND Token Metadata `save_v1_asset` into serializable transactions.  It might even be best to put **all** access of those tables into serializable transactions.  Either way it seems like this will lead to performance impact.
* Have to handle serialization failures and retry, as it won't actually lock.

### Traditional locking
#### Pros:
* Fine-grain control over the locking, not counting on PostgreSQL predicate locking/overall management.
* Might be more efficient than serializable transctions since there is only one place we need to lock.  This should prevent any deadlocks.

#### Cons:
* Have to manage locks in the code, which is error prone and could lead to deadlocks if done wrong.
* Won't provide the same level of isolation as serializable transactions.


### ChatGPT answer when I asked how to choose
* Choose serializable transactions when strong isolation guarantees are required, and you're willing to accept potential performance overhead.
* Choose explicit locks when you need fine-grained control over locking, and you're comfortable managing potential deadlocks and other complexities in your application code.
* In many cases, a combination of both approaches might be appropriate.  Performance testing is also recommended.

## Proposal
I think the locking requirements are minimal and only needed on the compression indexing side.  I am proposing an exclusive lock on the `asset_creators` table within a normal Read Committed isolation level transaction.  Alternatively I could also lock the `asset` table row but based on some analysis I do not think a second explicit lock is needed to protect against concurrency issues on `asset_creators`.

## Analysis
### Example database transactions
#### TX A (Bubblegum `mintV1`):
* lock `asset_creators` table
* check `asset.seq != 0`
* check `current seq > asset.base_info_seq `
* upsert base asset data
* delete creators with ID
* add creators with ID
* set verify flag if `seq > asset_creators.verify_seq` (this must be separate to support creator verify/unverify with higher `seq`)

#### TX B (Bubblegum `update_metadata`)
* lock `asset_creators` table
* check `asset.seq != 0`
* check `current seq > asset.base_info_seq`
* upsert base asset data
* delete creators with ID
* add creators with ID
* set verify flag if `seq > asset_creators.verify_seq`

#### TX C (Bubblegum `verify_creator`)
* lock `asset_creators` table
* check `asset.seq != 0`
* check `current seq > asset.base_info_seq`
* set verify flag if `seq > asset_creators.verify_seq`

#### TX D1 (Token metadata `save_v1_asset` to `asset` table):
* set `asset.seq` to 0
* set all other asset data

#### TX D2 (Token metadata `save_v1_asset` to `asset_creators` table):
* delete creators with ID
* add creators with ID including verified flag

### Concurrency cases
#### TX A and TX B are indexed concurrently
* One will block the other based on the lock to the `asset_creators` table.
* There will be no conflicts.  `asset.base_info_seq` will ensure proper ordering.

#### TX A and TX C are indexed concurrently
* One will block the other based on the lock to the `asset_creators` table.
* There will be no conflicts.  `asset_creators.verify_seq` will ensure proper ordering.

#### TX A and TXs C are indexed concurrently
_Case 1:_
* TX A gets the lock for the `asset_creators` table.
* TX D1 sets `asset.seq` to 0.
* TX D2 attempts to update `asset_creators` but must wait for lock.
* TX A does the check on `asset.seq` and returns early, returning the lock on `asset_creators`.
* TX D2 updates `asset_creators`.

_Case 2:_
* TX A gets the lock for the `asset_creators` table.
* TX A checks that `asset.seq != 0` and continues.
* TX D1 sets `asset.seq` to 0.
* TX D2 attempts to update `asset_creators` but must wait for lock.
* TX A updates `asset_creators` based on an old value for `asset.seq`, then returns the lock.
* TX D2 updates `asset_creators`.

_Case 3:_
* TX D1 sets `asset.seq` to 0.
* TX D2 starts to update the `asset_creators` table, which will acquire a row-level lock.
* TX A attempts to gen an exclusive lock on `asset_creators` table but must wait.
* TX D2 completes its update and returns its lock.
* TX A gets a lock on `asset_creators`
* TX A does the check on `asset.seq` and returns early, returning the lock on `asset_creators` table.

Reply:
What if we add a placeholder row in assets_creator to represent no creators? It's ugly but it escapes all the complexity.

Reply:
So if I add a "no creator" row to the asset_creators table, I think the idea is I can keep track of a sequence number in that row, and use that to manage creator updates adding/deleting based on that sequence number, so for example if there was an update metadata with an empty creators array, I would be able to handle concurrent updates with another instruction that would otherwise try to insert a row into asset_creators

Reply:
Subquery way would look like:
```
DELETE * FROM asset_creators WHERE ID = <id> AND current_seq >= (SELECT  asset_creators.update_seq FROM asset_creators WHERE ID = <id> LIMIT 1)
```

Separate statements would look something like:
```
txn.start();

let update_seq = txn.execute(SELECT asset_creators.update_seq FROM asset_creators WHERE ID = <id> LIMIT 1);

txn.execute(DELETE * FROM asset where ID = <id> AND current_seq > update_seq);

txn.execute(INSERT INTO asset_creators VALUES (...) ... ON CONFLICT ... AND current_seq > update_seq);

txn.commit();
```
I don't think either of these would lock the table between the separate SQL statements, as they affect different rows.  So then whatever sequence number we read out of the "empty creator" row could be stale

Reply:
I think this requires essentially selecting a row from the asset_creators table to get the current "update metadata" level sequence number, then using that value to decide whether to make further deletions/insertions to the table.  I don't think this will work with concurrent updates because selecting the row is one SQL statement in the table, and modifying the table is a second SQL statement, and some concurrent process could modify the table in between those two operations

Reply:
### Refactored creator indexing
* Use new Blockbuster that always updates all creators and verification status.
* Remove deleting creators with lower sequence numbers as it would not work due to race conditions.
* Add concept of "empty" creator value to support Bubblegum empty creator arrays.
* Add filtering out of old creators or having no creators to DAS code.
* Also get `authority` and `tree_id` accounts from Bubblegum during `mint` and `update_metadata`.

Reply:
Two possible solutions:
1. No-lock solution (what is currently in PR, described in [comment above](https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/134#discussion_r1429889104))
2. `asset` table row lock solution that @NicolasPennie also thought of while we were talking.

### No-lock solution
#### Pros:
* Removing an extra search/delete and no locks needed.  Always just upsert creator rows based on `slot` or `seq` to protect ordering/concurrency.
#### Cons:
* Database can store some stale creator rows, which are filtered out by DAS API (throws away anything with lower `slot` or `seq` number) at the time of API asset query.

### `asset` table lock solution:
#### Pros:
* Data is accurate in database, no DAS filtering required.
* Only a row-level lock on asset table which is why it is superior to my previous asset_creators table lock solution.
#### Cons:
* Requires a lock, still need to have the delete of existing rows before updating to account for creator array changing sizes.
* Does NOT remove the need for all DAS API filtering, because to make update_metadata work properly we need to have a <no_creators> row with a seq  to keep track, which needs to be removed by the API asset query.

Reply:
Talked with @NicolasPennie and @linuskendall offline and selected the no-lock solution.

I have tested this solution with a `mint_v1`, `verify_creator`, `unverify_creator`, and `update_metadata` sequence, and observed the database was in expected state and `getAsset` returned the expected result at each step.

Reply:
Found filtering issue with `getAssetsByCreator` returning results for stale creators.  Updated PR to fix the filtering.  Retested it and observe that when a creator is removed with `update_metadata`,  that `getAsset` and `getAssetsByCreator` work as expected in my test case.

Reply:
Marking this as resolved after testing it works